### DISCUSSION TITLE
phase4 3.x

### DISCUSSION THREAD

**Discussion by phax** [phax/phase4#266] [Show and tell]
Hi all,
This thread is meant to keep you updated about the development of phase4 3.x and the changes performed there.
As you might have seen in the code, a couple of backwards-compatible changes were not performed yet, and saved for the next major release. The respective code parts were marked with `@ChangePhase4V3` annotation.
Additionally, of course all deprecated methods, marked for removal, will be removed.

All the performed changed will be listed at the respective migration page: https://github.com/phax/phase4/wiki/Migrations

Any comments are welcome

**Comment by beth-soptim** [comment]
Is is planned to migrate to JDK 17 or 21 (would be ok for us)?

**Reply by phax** [reply]
The basis will stay with Java 11, as there is no need from any base library. The major release is focused on cleansing and alignment.
Only the Spring-Boot example is on Java 17 and I don't think they will change before the next major release either.

For version 4.x there will be a poll I assume to see where the pain points are - many still rely on Java 8....

**Comment by ri4a** [comment]
Would love to move to Java 17 - but not before #124 is resolved.

**Reply by phax** [reply]
You can already today move to Java 17 (or even Java 21) without an issue. The effective workaround to #124 is to use CRL, which works without a problem to the best of my knowledge.

**Comment by Jef-VDD** [comment]
Hello @phax,

Do you have a timeline or release date planned for the release of phase4 3.x?

**Reply by phax** [reply]
No not really. There are a few issues I want to resolve for v3, as they might inccur changes to the API (or at least they are easier to implement with API changes). However, September is a conference month and October is a vacation month. Realistically beginning of November I'd say. However, I can always release a "beta1" version if you are waiting for something particular...

**Reply by Jef-VDD** [reply]
Hey Philip,

We are indeed waiting for [this](https://github.com/phax/phase4/discussions/265) issue in particular. A beta version would help us start working on our integration and testing process earlier. Once it's available in Maven Central, we can align our work with the updates. Thanks for considering this!

**Reply by phax** [reply]
Okay, than I will create a beta1 release asap

**Reply by phax** [reply]
@Jef-VDD https://github.com/phax/phase4/releases/tag/phase4-parent-pom-3.0.0-beta1 is available

**Comment by Main-Tomi** [comment]
Dear @phax ,

The version I used before was 2.7.5. and I meet some error after I  use beta1.
In the version of 2.7.5, the class **AS4XServletHandler** have attribute **m_aCryptoFactorySupplier** which we can reset **IAS4CryptoFactory** dynamically. 
In beta1, **AS4XServletHandler** only have a attribute(IAS4ServletRequestHandlerCustomizer).
So I want to know if there are other ways to dynamically modify the key repository？

Best Regards,
Tomi Lin

**Reply by phax** [reply]
Hi @Main-Tomi ,
The intention is to move all the customizing to a central place, because it was a mixture between settings in `AS4XServletHandler` and in `AS4RequestHandler`. The new version cleans this up, and puts all the customizations into one place. Therefore you also don't need a `Supplier` anymore, but can pass the instance directly.

See below some Java code snippet on how to customize the crypto factory per request.
```java
   final AS4XServletHandler aHdl = new AS4XServletHandler ();
    aHdl.setRequestHandlerCustomizer (new IAS4ServletRequestHandlerCustomizer ()
    {
      public void customizeBeforeHandling (final IRequestWebScopeWithoutResponse aRequestScope,
                                           final AS4UnifiedResponse aUnifiedResponse,
                                           final AS4RequestHandler aRequestHandler)
      {
        aRequestHandler.setCryptoFactory (new AS4CryptoFactoryInMemoryKeyStore (AppConfig.loadKeyStore ()
                                                                                         .getKeyStore (),
                                                                                AppConfig.getKeyStoreKeyAlias (),
                                                                                AppConfig.getKeyStoreKeyPasswordString (),
                                                                                AppConfig.loadAS4TrustStore ()
                                                                                         .getKeyStore ()));
      }

      public void customizeAfterHandling (final IRequestWebScopeWithoutResponse aRequestScope,
                                          final AS4UnifiedResponse aUnifiedResponse,
                                          final AS4RequestHandler aRequestHandler)
      {}
    });

    // HTTP POST only
    handlerRegistry ().registerHandler (EHttpMethod.POST, aHdl);
```

hth

**Reply by Main-Tomi** [reply]
Hi @phax,

Thanks for you help, I tested it pass, thanks.

Best Regards,
Tomi

**Comment by phax** [comment]
So, finally version 3.0.0 is ready and published. I hope you all like it and benefit from the improvements, As usual, let me know if you have issues or troubles upgrading. Please consult https://github.com/phax/phase4/wiki/Migrations before :)
